src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,6 +232,42 @@ typedef struct MPIDI_prequest {
     MPI_Datatype datatype;
 } MPIDI_prequest_t;
 
+/* Structures for partitioned pt2pt request */
+typedef struct MPIDIG_part_sreq {
+    MPIR_cc_t ready_cntr;
+} MPIDIG_part_sreq_t;
+
+typedef struct MPIDIG_part_rreq {
+    MPI_Aint sdata_size;        /* size of entire send data */
+
+    /* fields for posted/unexp list */
+    MPIR_Request *request;      /* pointer of part_req */
+    struct MPIDIG_part_rreq *prev, *next;
+} MPIDIG_part_rreq_t;
+
+typedef struct MPIDIG_part_request {
+    MPL_atomic_int_t status;    /* see MPIDIG_PART_REQ_INC_FETCH_STATUS */
+    MPIR_Request *peer_req_ptr;
+    union {
+        MPIDIG_part_sreq_t sreq;
+        MPIDIG_part_rreq_t rreq;
+    } u;
+} MPIDIG_part_request_t;
+
+typedef struct MPIDI_part_request {
+    MPIDIG_part_request_t am;
+
+    /* partitioned attributes */
+    void *buffer;
+    int partitions;
+    MPI_Aint count;             /* count per partition */
+    int rank;
+    int tag;
+    MPIR_Context_id_t context_id;
+    MPI_Datatype datatype;
+    int active_flag;            /* flag indicating whether in a start-complete active period */
+} MPIDI_part_request_t;
```

Comment:
Usually, the request either go thru native path where use the MPIDI field, or go thru the AM path, where we use the am fields. Your design seems to require both the native part and am part, is this necessary? If possible, try not to create new patterns.

Reply:
I think we need both am and device-layer (MPIDI) fields. My reason is that, the partitioned req is more similar to an RMA window, rather than another type of `one-time-used-request`. Specifically,
- We need store a set of persistent attributes such as `buffer`, `partitions`, `count`, for many rounds of communication, similar to RMA mode. These attributes will be used for both AM path and native path. Thus, we put them in `MPIDI` as device-layer info.
- The AM part is used for MPIDIG am protocol. When we develop the native path in future, if we don't use the same protocol for start-complete sync, we may put `am` into a union at that time; likely we may want to simply reuse the AM protocol for sync as they are not in perf-critical path, similar to that of RMA (e.g., AM based sync + RDMA based perf-critical data transfer)

Reply:
My understanding is that you are OK if I keep the current request structure and keep all partitioned attributes in `MPIDI_part_request_t`. Can you please confirm?

Reply:
Yes, you can keep it for now. I dont' think it is critical. We can easily move the struct to different place if we want to, right?